realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/RealmFunctionsTests.kt

```diff
@@ -0,0 +1,150 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm
+
+import androidx.test.platform.app.InstrumentationRegistry
+import org.bson.*
+import org.bson.codecs.StringCodec
+import org.bson.codecs.configuration.CodecRegistries
+import org.bson.codecs.pojo.PojoCodecProvider
+import org.bson.types.Decimal128
+import org.bson.types.ObjectId
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import kotlin.test.assertEquals
+
+class RealmFunctionsTests {
+
+    private lateinit var app: TestRealmApp
+    private lateinit var functions : RealmFunctions
+
+    @Before
+    fun setup() {
+        Realm.init(InstrumentationRegistry.getInstrumentation().targetContext)
+        app = TestRealmApp()
+        functions = RealmFunctions(app.configuration.codecRegistry)
+    }
+
+    @After
+    fun teardown() {
+        if (this::app.isInitialized) {
+            app.close()
+        }
+    }
+
+    // Test of BSON JNI round trip until superseded with actual public api tests are added.
+    @Test
+    fun jniRoundTripForDefaultCodecRegistry() {
+        val i32 = 42
+        val i64 = 42L
+
+        for (type in BsonType.values()) {
+            when (type) {
+                BsonType.DOUBLE -> {
+                    assertTypedEcho(java.lang.Float(1.4), java.lang.Float::class.java)
+                    assertTypedEcho(java.lang.Double(1.4), java.lang.Double::class.java)
+                    assertTypedEcho(BsonDouble(1.4), BsonDouble::class.java)
+                }
+                BsonType.STRING -> {
+                    assertTypedEcho("Realm", String::class.java)
+                    assertTypedEcho(BsonString("Realm"), BsonString::class.java)
+                }
+                BsonType.ARRAY -> {
+                    // FIXME Fails in C++ parsing when boolean values are added...needs investigation
+                    //  io.realm.exceptions.RealmError: Unrecoverable error. current state '$1' is not of expected state '$2' in /Users/claus.rorbech/proj/realm-java/realm/realm-library/src/main/cpp/io_realm_RealmFunctions.cpp line 32
+                    //val listValues = listOf<Any>(true, i32, i64)
+                    val listValues = listOf<Any>(i32, i64)
+                    assertTypedEcho(listValues, List::class.java)
+                }
+                BsonType.BINARY -> {
+                    val value = byteArrayOf(1, 2, 3)
+                    val actual = functions.invoke(value, ByteArray::class.java)
+                    assertEquals(value.toList(), actual.toList())
+                    // FIXME Does not seem to preserve type
+                    // assertTypedEcho(BsonBinary(UUID.randomUUID()), BsonBinary::class.java)
```

Comment:
UUID is a custom type. Maybe because a CodeRegistry is missing?

Reply:
Seems to be that the C++ roundtrip actually reset subtube. I will raise it with OS.